### DISCUSSION TITLE
UPNP/NAT-PMP/STUN support

### DISCUSSION THREAD

**Discussion by skbeh** [rustdesk/rustdesk#2730] [Ideas]
All of the three protocols in the title help with NAT traversal to some extent.
If the support is built-in, establishing direct connection will be much easier and the load of proxy server will be reduced.
Also, if such features are built-in, we will not have to set up private network to establish direct connection. In most situations, setting up private network is not even possible, such as remote controlling a random PC temporarily.

**Comment by TCB13** [comment]
+1 on this. Most remote solutions go for UPNP and it they can't reach the peers then they fallback to some kind of rendezvous server.

The current implementation works well but without a self-hosted server connections are slow the proxies are forced to handle a TON of unnecessary traffic.

**Comment by huangsijun17** [comment]
If UPnP is supported, it is very convenient to directly connect to the remote computer, saving a lot of traffic burden on the server. However, it is desirable to encrypt the traffic when handshaking with the server.

**Comment by bhack** [comment]
https://crates.io/crates/igd-next

**Comment by lurenjia1213** [comment]
So, it looks like rustdesk still doesn't support upnp at the moment?
